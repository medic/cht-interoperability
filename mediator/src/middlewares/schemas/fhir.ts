import { Fhir } from "fhir";

const fhir = new Fhir();

export function validateFhirResource(resourceType: string) {
  return function wrapper(data: any) {
    return fhir.validate({...data, resourceType})
  }
}
