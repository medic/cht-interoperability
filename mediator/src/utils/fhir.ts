export const VALID_GENDERS = ['male', 'female', 'other', 'unknown'] as const;

export interface IPatient {
  name: string;
  date_of_birth: string;
  _id: string;
  sex: typeof VALID_GENDERS[number];
}

export function generateFHIRPatientResource(patient: IPatient) {
  if (!patient.name) {
    throw new Error(`Invalid 'name' expected type of 'string' but recieved '${patient.name}'`);
  }
  
  const patientLastName = patient.name.split(' ').slice(-1);
  const birthDate = new Date(patient.date_of_birth);

  if (!isValidDate(birthDate)) {
    throw new RangeError("Invalid 'date_of_birth' range: received " + patient.date_of_birth);
  } else if (!patient._id) {
    throw new Error(`Invalid '_id' epxted type of 'string' or 'number' but recieved '${typeof patient._id}' with value '${patient._id}'`);
  }  else if (!patient.sex || !VALID_GENDERS.includes(patient.sex)) {
    throw new Error(`Invalid 'sex' expected 'male', 'female', 'other', 'unknown' but recieved '${patient.sex}'`);
  }

  const FHITPatientResource = {
    resourceType: 'Patient',
    id: patient._id,
    identifier: [
      {
        system: 'cht',
        value: patient._id
      }
    ],
    name: [
      {
        use: 'official',
        family: patientLastName,
        given: [patient.name]
      }
    ],
    gender: patient.sex,
    birthDate: birthDate.toISOString()
  };

  return FHITPatientResource;
}

export function isValidDate(d: Date) {
  return d instanceof Date && !isNaN(d as any);
}

export function generateFHIRSubscriptionResource(patientId: string, callbackUrl: string) {
  if (!patientId) {
    throw new Error(`Invalid patient id was expecting type of 'string' or 'number' but received '${typeof patientId}'`)
  } else if (!callbackUrl) {
    throw new Error(`Invalid 'callbackUrl' was expecting type of 'string' but recieved '${typeof callbackUrl}'`)
  }
  
  const FHIRSubscriptionResource = {
    resourceType: 'Subscription',
    id: patientId,
    status: 'requested',
    reason: 'Follow up request for patient',
    criteria: `Encounter?identifier=${patientId}`,
    channel: {
      type: 'rest-hook',
      endpoint: callbackUrl,
      payload: 'application/fhir+json',
      header: ['Content-Type: application/fhir+json']
    }
  };

  return FHIRSubscriptionResource;
}
