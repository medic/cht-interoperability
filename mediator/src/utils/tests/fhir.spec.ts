import { generateFHIRPatientResource, generateFHIRSubscriptionResource, IPatient, isValidDate } from "../fhir";

describe("generateFHIRPatientResource", () => {
  const validChtPatient: IPatient = {
    _id: "456",
    name: "John Doe",
    sex: "male",
    date_of_birth: "2000-01-01",
  };

  it("creates a valid FHIR Patient resource when all the input are available", () => {
    const patient = generateFHIRPatientResource({ ...validChtPatient });

    expect(patient).toEqual({
      resourceType: "Patient",
      id: "456",
      identifier: [
        {
          system: "cht",
          value: "456",
        },
      ],
      name: [
        {
          use: "official",
          family: ["Doe"],
          given: ["John Doe"],
        },
      ],
      gender: "male",
      birthDate: "2000-01-01T00:00:00.000Z",
    });
  });

  it("throws an error if the 'date_of_birth' is 'undefined'", () => {
    const invalidChtPatient = { ...validChtPatient, date_of_birth: undefined } as any;

    expect(() =>
      generateFHIRPatientResource(invalidChtPatient)
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws an error when 'date_of_birth' is an invalid date", () => {
    const invalidChtPatient = {
      ...validChtPatient,
      date_of_birth: "YYYY-MM-DD",
    };

    expect(() =>
      generateFHIRPatientResource(invalidChtPatient)
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws an error when 'name' is an invalid name", () => {
    const invalidChtPatient: any = {
      ...validChtPatient,
      name: undefined,
    };

    expect(() =>
      generateFHIRPatientResource(invalidChtPatient)
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws an error when '_id' is an invalid id", () => {
    const invalidChtPatient: any = {
      ...validChtPatient,
      _id: undefined,
    };

    expect(() =>
      generateFHIRPatientResource(invalidChtPatient)
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws an error when 'sex' is an invalid sex", () => {
    const invalidChtPatient: any = {
      ...validChtPatient,
      sex: 'INVALID_GENDER',
    };

    expect(() =>
      generateFHIRPatientResource(invalidChtPatient)
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws an error when 'sex' is undefined", () => {
    const invalidChtPatient: any = {
      ...validChtPatient,
      sex: undefined,
    };

    expect(() =>
      generateFHIRPatientResource(invalidChtPatient)
    ).toThrowErrorMatchingSnapshot();
  });
});

describe("generateFHIRSubscriptionResource", () => {
  const patientId = "PATIENT_ID"
  const callbackUrl = "CALLBACK_URL"

  it("generates a subscription resource when passed valid 'patientId' and 'callbackUrl'", () => {
    const resource = generateFHIRSubscriptionResource(patientId, callbackUrl);

    expect(resource).toEqual({
      resourceType: 'Subscription',
      id: patientId,
      status: 'requested',
      reason: 'Follow up request for patient',
      criteria: `Encounter?identifier=${patientId}`,
      channel: {
        type: 'rest-hook',
        endpoint: callbackUrl,
        payload: 'application/fhir+json',
        header: ['Content-Type: application/fhir+json']
      }
    })
  })

  it("fails to generate subscription when given an invalid 'patientId'", () => {
    expect(() => generateFHIRSubscriptionResource(undefined as any, callbackUrl))
      .toThrowErrorMatchingSnapshot()
  })

  it("fails to generate subscription when given an invalid 'callbackUrl'", () => {
    expect(() => generateFHIRSubscriptionResource(patientId, undefined as any))
      .toThrowErrorMatchingSnapshot()
  })
})

describe("isValidDate", () => {
  it("returns true for a valid date", () => {
    expect(isValidDate(new Date("01-01-2001"))).toBe(true);
  })

  it("return false for an invalid date", () => {
    expect(isValidDate(new Date(NaN))).toBe(false);
  })

  it("return false for a none date object", () => {
    expect(isValidDate({} as any)).toBe(false);
  })
})