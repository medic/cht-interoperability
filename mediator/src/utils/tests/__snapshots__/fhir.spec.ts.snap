// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateFHIRPatientResource throws an error if the 'date_of_birth' is 'undefined' 1`] = `"Invalid 'date_of_birth' range: received undefined"`;

exports[`generateFHIRPatientResource throws an error when '_id' is an invalid id 1`] = `"Invalid '_id' epxted type of 'string' or 'number' but recieved 'undefined' with value 'undefined'"`;

exports[`generateFHIRPatientResource throws an error when 'date_of_birth' is an invalid date 1`] = `"Invalid 'date_of_birth' range: received YYYY-MM-DD"`;

exports[`generateFHIRPatientResource throws an error when 'name' is an invalid name 1`] = `"Invalid 'name' expected type of 'string' but recieved 'undefined'"`;

exports[`generateFHIRPatientResource throws an error when 'sex' is an invalid sex 1`] = `"Invalid 'sex' expected 'male', 'female', 'other', 'unknown' but recieved 'INVALID_GENDER'"`;

exports[`generateFHIRPatientResource throws an error when 'sex' is undefined 1`] = `"Invalid 'sex' expected 'male', 'female', 'other', 'unknown' but recieved 'undefined'"`;

exports[`generateFHIRSubscriptionResource fails to generate subscription when given an invalid 'callbackUrl' 1`] = `"Invalid 'callbackUrl' was expecting type of 'string' but recieved 'undefined'"`;

exports[`generateFHIRSubscriptionResource fails to generate subscription when given an invalid 'patientId' 1`] = `"Invalid patient id was expecting type of 'string' or 'number' but received 'undefined'"`;
